File: Alam's/agentra_travelagent/lib/models/agent.dart
import '../config/api_config.dart';

class Agent {
  final String id;
  final String fullName;
  final String businessName;
  final String email;
  final String phone;
  final String cnic;
  final String? profileImage;
  final String? location;
  final String? refundPolicy;
  final String? cancellationPolicy;
  final String role;
  final bool isVerified;
  final String status;
  final String? rejectionReason;
  final int totalPackages;
  final int totalBookings;
  final double averageRating;

  // AI Subscription fields
  final String subscriptionPlan; // FREE | MONTHLY | YEARLY
  final bool subscriptionActive;
  final DateTime? subscriptionExpiry;

  Agent({
    required this.id,
    required this.fullName,
    required this.businessName,
    required this.email,
    required this.phone,
    required this.cnic,
    this.profileImage,
    this.location,
    this.refundPolicy,
    this.cancellationPolicy,
    this.role = 'AGENT',
    this.isVerified = false,
    this.status = 'PENDING_APPROVAL',
    this.rejectionReason,
    this.totalPackages = 0,
    this.totalBookings = 0,
    this.averageRating = 0.0,
    this.subscriptionPlan = 'FREE',
    this.subscriptionActive = false,
    this.subscriptionExpiry,
  });

  bool get isPro =>
      subscriptionActive &&
      (subscriptionPlan == 'MONTHLY' || subscriptionPlan == 'YEARLY') &&
      (subscriptionExpiry == null ||
          subscriptionExpiry!.isAfter(DateTime.now()));

  factory Agent.fromJson(Map<String, dynamic> json) {
    final aiSub = json['aiSubscription'] as Map<String, dynamic>? ?? {};
    return Agent(
      id: json['_id'] ?? '',
      fullName: json['fullName'] ?? '',
      businessName: json['businessName'] ?? '',
      email: json['email'] ?? '',
      phone: json['phone'] ?? '',
      cnic: json['cnic'] ?? '',
      profileImage: ApiConfig.getImageUrl(json['profileImage']),
      location: json['location'],
      refundPolicy: json['refundPolicy'],
      cancellationPolicy: json['cancellationPolicy'],
      role: json['role'] ?? 'AGENT',
      isVerified: json['isVerified'] ?? false,
      status: json['status'] ?? 'PENDING_APPROVAL',
      rejectionReason: json['rejectionReason'],
      totalPackages: json['totalPackages'] ?? 0,
      totalBookings: (json['totalBookings'] ?? 0) < 0 ? 0 : (json['totalBookings'] ?? 0),
      averageRating: (json['averageRating'] ?? 0).toDouble(),
      subscriptionPlan: aiSub['plan'] ?? 'FREE',
      subscriptionActive: aiSub['isActive'] ?? false,
      subscriptionExpiry: aiSub['expiryDate'] != null
          ? DateTime.tryParse(aiSub['expiryDate'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'fullName': fullName,
      'businessName': businessName,
      'email': email,
      'phone': phone,
      'cnic': cnic,
      'profileImage': profileImage,
      'location': location,
      'refundPolicy': refundPolicy,
      'cancellationPolicy': cancellationPolicy,
    };
  }
}

File: Alam's/agentra-backend/src/models/Agent.js
const mongoose = require('mongoose');

const agentSchema = new mongoose.Schema({

  // Basic Info
  fullName: {
    type: String,
    required: true,
    trim: true
  },

  businessName: {
    type: String,
    required: true,
    trim: true
  },

  email: {
    type: String,
    required: true,
    unique: true,
    lowercase: true
  },

  password: {
    type: String,
    required: true
  },

  phone: {
    type: String,
    required: true,
    unique: true
  },

  cnic: {
    type: String,
    required: true,
    unique: true
  },

  location: {
    type: String,
    default: ''
  },

  bio: {
    type: String,
    default: ''
  },

  profileImage: {
    type: String,
    default: ''
  },

  // Policies
  refundPolicy: {
    type: String,
    default: ''
  },

  cancellationPolicy: {
    type: String,
    default: ''
  },

  // System Control
  role: {
    type: String,
    enum: ['AGENT'],
    default: 'AGENT'
  },

  // Approval Status
  status: {
    type: String,
    enum: ['PENDING_APPROVAL', 'APPROVED', 'REJECTED', 'BLOCKED'],
    default: 'PENDING_APPROVAL'
  },

  rejectionReason: {
    type: String,
    default: ''
  },

  isVerified: {
    type: Boolean,
    default: false
  },

  verificationDocuments: [{
    type: String
  }],

  emailVerified: {
    type: Boolean,
    default: false
  },

  emailVerificationToken: {
    type: String
  },

  resetPasswordToken: {
    type: String
  },

  resetPasswordExpires: {
    type: Date
  },

  // ── owner Notice / Warning ──────────────────────────────────────
  // Set when owner sends a 30-day notice before blocking or deleting
  noticeSentAt: {
    type: Date,
    default: null
  },

  noticeType: {
    type: String,
    enum: ['block', 'delete', null],
    default: null
  },

  noticeMessage: {
    type: String,
    default: ''
  },

  // Business Stats (used in dashboards)
  totalPackages: {
    type: Number,
    default: 0
  },

  totalBookings: {
    type: Number,
    default: 0
  },

  averageRating: {
    type: Number,
    default: 0
  },

  // AI Tools & Subscription
  aiSubscription: {
    plan: { type: String, enum: ['FREE', 'MONTHLY', 'YEARLY'], default: 'FREE' },
    isActive: { type: Boolean, default: false },
    expiryDate: Date
  }

}, { timestamps: true });

module.exports = mongoose.model('Agent', agentSchema);
